tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java

```diff
@@ -138,14 +217,101 @@ public GraphOperation operation(String name) {
     return new OperationIterator(this);
   }
 
+  private GraphOperation graphOp(Operand<?> operand) {
+    checkInput(operand);
+    return (GraphOperation) operand.op();
+  }
+
+  /**
+   * Finds the operations used to produce {@code outputs} from {@code inputs}, or throws if that is not possible.
+   * Respects control dependencies.
+   *
+   * If both {@code allowedBodyOps} and {@code forbiddenBodyOps} are null, forbids 0-input ops in the body.
+   *
+   * @param inputs the inputs of the subgraph.  Must be from single output ops.
+   * @param outputs the outputs of the subgraph
+   * @param allowedBodyOps types of ops to allow as 0-input ops in the body.  Allows all (except {@code
+   * forbiddenBodyOps}) if null.
+   * @param forbiddenBodyOps types of ops to never allow as 0-input ops in the body.  Forbids all (except {@code
+   * allowedBodyOps}) if null.
+   * @return the set of operations needed to calculate outputs from inputs, including outputs and inputs
+   * @throws IllegalStateException if outputs depends on ops outside of the subgraph (i.e. is not calculable based
+   * solely on inputs)
+   */
+  public synchronized Set<GraphOperation> completeSubgraph(Set<Operand<?>> inputs, Set<Operand<?>> outputs,
+      Set<String> allowedBodyOps, Set<String> forbiddenBodyOps) {
+
+    if (forbiddenBodyOps != null && allowedBodyOps != null) {
+      for (String t : forbiddenBodyOps) {
+        if (allowedBodyOps.contains(t)) {
+          throw new IllegalArgumentException("Can't allow and forbid op type " + t + ".");
+        }
+      }
+    }
+
+    Queue<GraphOperation> currents = new LinkedList<>();
+    Set<GraphOperation> seen = new LinkedHashSet<>(outputs.size());
```

Comment:
`seen` is allocated as `outputs.size()`, but we immediately add each element of `inputs` to it. Should it be sized as `inputs.size()` (or even `inputs.size() + outputs.size()`)?